# Thông tin đồ án

## Thành viên

|   MSSV   |      Họ tên     |   GitHub  |
|:--------:|:---------------:|:---------:|
| 18120120 | Trần Chí Hào    | haotrch   |
| 18120147 | Trần Quốc Thắng | tqthangbl |

## Link thùng chứa GitHub

[https://github.com/tqthangbl/P4DS-FinalProject](https://github.com/tqthangbl/P4DS-FinalProject)

# Môi trường code



- Code trên môi trường Linux
- Dùng phiên bản các package như trong file "min_ds-env.yml"


In [1]:
import sys
sys.executable

'/home/family-u/.conda/envs/min_ds-env/bin/python'

# Import các thư viện cần thiết

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Thu thập dữ liệu

**Dữ liệu của bạn là về chủ đề gì và bạn lấy từ nguồn nào?**

Bộ dữ liệu mà nhóm sử dụng có tên là "Google Play Store Apps", được lấy từ [Kaggle](https://www.kaggle.com/lava18/google-play-store-apps), chứa thông tin về hơn 10000 ứng dụng trên Google Play Store. Thời điểm bộ dữ liệu được đăng lên Kaggle là 2018-09-05, lần cập nhật cuối là vào 2019-02-03.

Có 2 file dữ liệu:
1. `googleplaystore.csv`: Chứa các thông tin chi tiết về các ứng dụng thu thập được từ Google Play
2. `googleplaystore_user_reviews`: Chứa thông tin về 100 đánh giá liên quan nhất cho mỗi ứng dụng thu thập được từ Google Play

**Người ta có cho phép bạn dùng dữ liệu như này không? Ví dụ, bạn có thể xem thử license của dữ liệu là gì**


License của dữ liệu là Creative Commons Attribution 3.0 Unported, theo như được nói ở [đây](https://creativecommons.org/licenses/by/3.0/), người ta cho phép:

1. Copy và chia sẻ dữ liệu này dưới bất kì hình thức nào
2. Sử dụng bộ dữ liệu này để xây dựng bất cứ thứ gì, cho bất cứ mục đích nào, kể cả thương mại

**Người ta thu thập dữ liệu như thế nào?**

Dữ liệu được tác giả crawl từ Google Play Store.


# Khám phá dữ liệu (thường đan xen với tiền xử lý)

* Dữ liệu có bao nhiêu dòng và bao nhiêu cột?
* Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa
khác nhau không?
* Dữ liệu có các dòng bị lặp không?
* Mỗi cột có ý nghĩa gì?
* Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu
dữ liệu chưa phù hợp để có thể xử lý tiếp không?
* Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá
trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Min? max? → Có gì bất thường không?
* Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical),
các giá trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Số lượng các giá trị khác nhau? Show một vài giá trị
→ Có gì bất thường không?

## Đọc dữ liệu từ file csv vào dataframe

Ở đây nhóm khám phá dữ liệu trong file `googleplaystore.csv` trước. Dữ liệu đọc từ file sẽ lưu vào dataframe `app_df`.

In [3]:
app_df = pd.read_csv('Data/googleplaystore.csv')
app_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

## Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
num_rows = len(app_df)
num_cols = len(app_df.columns)

print(f'[+] num_rows = {num_rows}')
print(f'[+] num_cols = {num_cols}')

[+] num_rows = 10841
[+] num_cols = 13


Dữ liệu có 10841 dòng và 13 cột.

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [5]:
app_df.sample(5)

App        Category  Rating Reviews  \
997                    Funny Quotes Free   ENTERTAINMENT     4.2   23666   
1466                  Redfin Real Estate  HOUSE_AND_HOME     4.6   36857   
7383                      Thistletown CI    PRODUCTIVITY     1.0       1   
5363                       I Am Innocent            GAME     4.4  208501   
4826  Zombie War Z : Hero Survival Rules            GAME     3.9    1987   

      Size    Installs  Type Price Content Rating         Genres  \
997   3.6M  1,000,000+  Free     0           Teen  Entertainment   
1466   19M  1,000,000+  Free     0       Everyone   House & Home   
7383  6.6M        100+  Free     0       Everyone   Productivity   
5363   56M  1,000,000+  Free     0           Teen      Adventure   
4826   62M    100,000+  Free     0     Mature 17+         Action   

           Last Updated Current Ver   Android Ver  
997         May 4, 2018         8.1    4.0 and up  
1466      July 25, 2018       220.0    5.0 and up  
7383     March 15, 2018        41.9    4.1 and up  
5363      July 23, 2018     2.8.307  4.0.3 and up  
4826  December 13, 2017         1.8  4.0.3 and up

Theo như mô tả của tác giả, mỗi dòng trong dữ liệu chứa thông tin của 1 ứng dụng trên Play Store. Bằng việc quan sát ngẫu nhiên các dòng trong dữ liệu nhiều lần, có vẻ không có dòng nào mang ý nghĩa khác. 

## Dữ liệu có các dòng bị lặp không?

In [10]:
have_duplicated_rows = True
dup_df = app_df[app_df.duplicated()]
if len(dup_df) == 0:
    have_duplicated_rows = False
    
print(f'[+] have_duplicated_rows = {have_duplicated_rows}')

[+] have_duplicated_rows = True


### Vấn đề cần tiền xử lý
- Một số dòng trong dữ liệu bị lặp, điều này có thể dẫn đến khi tính toán (đếm, tính tỉ lệ). Do đó cần phải loại bỏ những dòng bị lặp.

---

# Tiền xử lý

## Loại bỏ các dòng bị lặp

Các dòng bị lặp được lưu trong dataframe `dup_df`. Đầu tiên xem thử 1 vài giá trị của dataframe này.

In [17]:
dup_df.sample(5)

App            Category  \
3811                     CNN Breaking US & World News  NEWS_AND_MAGAZINES   
2645                                            LOVOO              SOCIAL   
2778                      JackThreads: Men's Shopping            SHOPPING   
2745  Shopkick: Free Gift Cards, Shop Rewards & Deals            SHOPPING   
474                       LINE: Free Calls & Messages       COMMUNICATION   

      Rating   Reviews                Size      Installs  Type Price  \
3811     4.0    293080                 25M   10,000,000+  Free     0   
2645     4.0    852455                 68M   10,000,000+  Free     0   
2778     4.0     13085                2.7M    1,000,000+  Free     0   
2745     4.3    213735                 43M   10,000,000+  Free     0   
474      4.2  10790289  Varies with device  500,000,000+  Free     0   

     Content Rating            Genres      Last Updated         Current Ver  \
3811   Everyone 10+  News & Magazines    August 6, 2018                5.17   
2645     Mature 17+            Social    August 3, 2018                27.0   
2778           Teen          Shopping  November 8, 2017                 5.1   
2745       Everyone          Shopping     July 24, 2018              5.3.73   
474        Everyone     Communication     July 26, 2018  Varies with device   

             Android Ver  
3811          4.4 and up  
2645          4.2 and up  
2778        4.0.3 and up  
2745          4.4 and up  
474   Varies with device

Tiến hành loại bỏ các dòng bị lặp trong `app_df` và cập nhật lại biến `num_rows` thể hiện số dòng của dữ liệu.

In [21]:
app_df = app_df.drop_duplicates()
num_rows = len(app_df)

print(f'[+] num_rows = {num_rows}')

[+] num_rows = 10358


# Đưa ra các câu hỏi có ý nghĩa cần trả lời

* Câu hỏi là gì?
* Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

# Nhìn lại quá trình làm đồ án

* Mỗi thành viên: Đã gặp những khó khăn gì? (Hay mọi
chuyện đều thuận lợi)
* Mỗi thành viên: Có học được gì hữu ích? (Hay không
học được gì)
* Nhóm: Nếu có thêm thời gian thì sẽ làm gì?

# Tài liệu tham khảo